### PR TITLE
Fix zero length types

### PR DESCRIPTION
**Pull request by eun-ice**
This pull request seems to fix a memory issue, by just referring to Box::new for zero-length types.

Since Box does not allocate for zero-length types and mem::size_of::<T> is constant for any given generic type, this change should not negatively affect performance.

### CODE REVIEW

File: src/boxed.rs [hunk 1]
```diff
@@ -1,12 +1,15 @@
 use std::{alloc, mem, ptr};
 
-
 /// A typesafe helper that stores the allocated pointer without the data initialized.
 pub struct BoxAllocation<T>(*mut T);
 
 impl<T> BoxAllocation<T> {
     /// Consumes self and writes the given value into the allocation.
     pub fn init(mut self, value: T) -> Box<T> {
+        if ::std::mem::size_of::<T>() == 0 {
```
Comment (kvark) [inline src/boxed.rs]: Let's use the already imported `mem`

**Comment by eun-ice** [comment]
> how does Box handle that? Maybe we can just do the same here instead of checking for size twice and returning Box::new(value)
>have you confirmed that the extra statements don't ruin our optimized LLVM IR?

The size check is zero-cost at runtime with optimization on. Rust generates binary code for every <T>, so mem::size_of::<T>() can be replaced with a constant. 

In fact: The entire call is removed by the optimizer:

Check the assembly here:

https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=67f10928f318f12a6f443df1b3a81ca4

```rust
        Box::alloc().init(EmptyType{})
```
becomes
```asm
	movl	$1, %ebx
```

while 

```rust
        Box::alloc().init(NonEmptyType{ string:"some string".into()})
```

starts with a call to __rust_alloc right away, without any checks:

``` asm
	movq	__rust_alloc@GOTPCREL(%rip), %r14
	movl	$24, %edi
	movl	$8, %esi
	callq	*%r14
```

So my PR should not have any consequences on runtime behaviour if optimization is turned on - except for being faster with empty types

> Notes: this conflicts with #1, which ideally should go first

That is your decision. If you merge #1 I will gladly adapt this PR. But IMO a memory safety error is of higher priority.

File: src/boxed.rs (hunk 1 above)
Comment (eun-ice) [reply src/boxed.rs]: PR updated as per your suggestion.

**Comment by kvark** [comment]
Ok, fair enough. Thank you!
bors r+

**Pull request merged**